server/modlog.ts

```diff
@@ -171,7 +171,9 @@ export class Modlog {
 	async rename(oldID: ModlogID, newID: ModlogID) {
 		const streamExists = this.streams.has(oldID);
 		if (streamExists) await this.destroy(oldID);
-		await FS(`${this.logPath}/modlog_${oldID}.txt`).rename(`${this.logPath}/modlog_${newID}.txt`);
+		if (!this.getSharedID(oldID)) {
+			await FS(`${this.logPath}/modlog_${oldID}.txt`).rename(`${this.logPath}/modlog_${newID}.txt`);
+		}
 		if (streamExists) this.initialize(newID);
```

Comment:
So, the problem is that this function races with `this.write`.

Look at `this.write`:

```
	write(roomid: ModlogID, message: string, overrideID?: string) {
		const stream = this.streams.get(roomid);
		if (!stream) throw new Error(`Attempted to write to an uninitialized modlog stream for the room '${roomid}'`);
		void stream.write(`[${new Date().toJSON()}] (${overrideID || roomid}) ${message}\n`);
	}
```

It will crash if it's called during either of the two `await`s in this rename function.

To be fair, this isn't a problem your PR introduces, and propagating the Promise will not fix this problem at all... but it does need to be fixed at some point.

Reply:
If renames are atomic (which I believe they nearly always are), we could initialize the new stream _immediately_ instead of waiting. So `void` instead of `await` in every call here. I believe that's sufficient to solve the race condition (except in obscure situations where the rename isn't atomic).

If we can't assume that the rename is atomic or want to do this the "right" way, you would want to store a promise for the stream that `write` can use to wait until it can write for real.

Reply:
> it does need to be fixed at some point.

We're migrating modlogs to SQLite, which should solve the race condition.